File: 2022/day01/haskell/Main.hs
module Main where
import Data.List (sort)

group :: [String] -> [[String]]
group [] = []
group xs = takeWhile (/= "") xs : group (drop 1 $ dropWhile (/= "") xs)

main :: IO ()
main = do
  input <- lines <$> readFile "../input.txt"

  let sums = map (sum . (map read :: [String] -> [Int])) $ group input

  let max = maximum sums
  let top_3 = sum $ take 3 $ reverse $ sort sums

  putStrLn $ "Part 1: " ++ show max
  putStrLn $ "Part 2: " ++ show top_3

File: 2022/day01/rust/src/main.rs
use std::fs;

fn main() {
    let content =
        fs::read_to_string("../input.txt").expect("Something went wrong reading the file");
    let mut elves: Vec<i32> = vec![];
    let lines = content.lines();

    let mut tmp = vec![];
    for line in lines {
        if line.is_empty() {
            elves.push(
                tmp.into_iter()
                    .map(|x: String| x.parse::<i32>().expect("could not parse"))
                    .sum(),
            );
            tmp = vec![];
        } else {
            tmp.push(line.to_string());
        }
    }
    elves.push(
        tmp.into_iter()
            .map(|x: String| x.parse::<i32>().expect("could not parse"))
            .sum(),
    );
    elves.sort();

    println!("Part 1: {}", elves.last().unwrap());
    println!("Part 2: {}", elves.iter().rev().take(3).sum::<i32>());
}
